### PR TITLE
[12.0][MIG] purchase_triple_discount

### PR DESCRIPTION
**Pull request by legalsylvain**
~Required PR : https://github.com/OCA/purchase-workflow/pull/731 (CI will fail as soon as this PR is not merged)~

- Port purchase_triple_discount in V12.

- remove dependency to obsolete module product_supplierinfo_discount (now there are only ``purchase_discount`` and ``purchase_triple_discount module``). The two modules ``product_supplierinfo_discount`` and ``product_supplierinfo_triple_discount`` are depecrated. 
- rename ``procurement.rule`` into ``stock.rule``
- Use new hook introduced in purchase_discount V12 to add ``discount2`` and ``discount3`` for the model ``purchase.order.line`` and ``stock.rule`` (https://github.com/solvosci/purchase-workflow/pull/1)
- Introduce discount2 and discount3 in purchase.report model, removing roadmap issue.

### CODE REVIEW

File: purchase_triple_discount/__manifest__.py [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 {
-    'name': 'Account Invoice Triple Discount',
-    'version': '11.0.1.0.0',
+    'name': 'Purchase Order Triple Discount',
+    'version': '12.0.1.0.0',
     'category': 'Purchase Management',
-    'author': 'Tecnativa, '
-              'Odoo Community Association (OCA)',
+    'author': 'Tecnativa,'
+            'GRAP,'
```
Comment (pedrobaeza) [inline purchase_triple_discount/__manifest__.py]: Align this correctly

File: purchase_triple_discount/models/product_supplierinfo.py [hunk 2]
```diff
@@ -40,4 +40,4 @@ def create(self, vals):
                 vals['product_tmpl_id'] in self.env.context['discount3_map']):
             vals['discount3'] = self.env.context['discount3_map'][
                 vals['product_tmpl_id']]
-        return super(ProductSupplierInfo, self).create(vals)
+        return super().create(vals)
```
Comment (pedrobaeza) [inline purchase_triple_discount/models/product_supplierinfo.py]: Maybe we should introduce here a hook in `purchase_discount` as well for not having to check the mapping again (and include all lines in the initial mapping)?

File: purchase_triple_discount/models/purchase_order.py [hunk 3]
```diff
@@ -19,15 +19,15 @@ def _add_supplier_to_product(self):
...
             discount2_map=discount2_map, discount3_map=discount3_map)
-            )._add_supplier_to_product()
+        )._add_supplier_to_product()
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = "purchase.order.line"
 
     @api.depends('discount2', 'discount3')
     def _compute_amount(self):
```
Comment (pedrobaeza) [inline purchase_triple_discount/models/purchase_order.py:12]: Add the comment for clarifying that this is only for changing depends of the computed

File: purchase_triple_discount/models/stock_rule.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from odoo import api, models
+
+
+class StockRule(models.Model):
+    _inherit = 'stock.rule'
+
+    @api.model
+    def _prepare_purchase_order_line_from_seller(self, seller):
+        """Overload this function to prepare other data from seller,
+        like in purchase_triple_discount module"""
```
Comment (pedrobaeza) [inline purchase_triple_discount/models/stock_rule.py]: Self-referencing?

File: purchase_triple_discount/report/purchase_report.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    def _get_discounted_price_unit_exp(self):
+        """Inheritable method for getting the SQL expression used for
+        calculating the unit price with discount(s).
+
+        :rtype: str
+        :return: SQL expression for discounted unit price.
+        """
+        return """
+            (1.0 - COALESCE(l.discount, 0.0) / 100.0) *
```
Comment (pedrobaeza) [inline purchase_triple_discount/report/purchase_report.py]: Can this be simplified somehow for better performance?

File: purchase_triple_discount/models/stock_rule.py (hunk 4 above)
Comment (legalsylvain) [reply purchase_triple_discount/models/stock_rule.py]: indeed ! bad copy past from ``purchase_discount`` module. Thanks.

File: purchase_triple_discount/report/purchase_report.py (hunk 5 above)
Comment (legalsylvain) [reply purchase_triple_discount/report/purchase_report.py]: What do you mean exactly ? 
In fact, I'm really newb in ``report`` features. (I don't use it, for GRAP, I only used materialized view with ``bi_sql_editor``). I wrote that part only to remove the roadmap part ;-)

is ``COALESCE`` requiring a lot of time to compute ? I don't see how to make it more simple.

File: purchase_triple_discount/models/product_supplierinfo.py (hunk 2 above)
Comment (legalsylvain) [reply purchase_triple_discount/models/product_supplierinfo.py]: Yes, thanks, good idea. What about this diff ?

https://github.com/OCA/purchase-workflow/pull/733/commits/0a670aa1547f4add0654f85478a8ebde8c7e8f05

regards.

File: purchase_triple_discount/report/purchase_report.py (hunk 5 above)
Comment (pedrobaeza) [reply purchase_triple_discount/report/purchase_report.py]: It was more the number of operations. I'm proposing a refactoring in short.

**Comment by pedrobaeza** [comment]
@legalsylvain I have touched directly the branch for not being with PRs to your branch and take the opportunity to do some squashing on useless previous version administrative commits (POT, icon and so on), and on your migration commits, and add following improvements:

* Lighter algorithm when synchronizing PO line info to product.supplierinfo.
* Use new create multi method for performance.
* Better SQL expresion in purchase report for decreasing the number of operations.
* Add test for purchase report.

I hope you don't mind that I have taken this path. If so, just tell me and I'll do it other way next time.

P.S.: It would desirable that you use `fixup!` syntax when adding new commits that are PR refinements for being auto-squashed on merge, or directly to amend previous commit, as GitHub already shows the diff between the previous commit and the current one.

**Comment by legalsylvain** [comment]
Hi @pedrobaeza 
first, thanks for your contribution.
In a genaral way, no problem to commit directly on my PR, if : 
- the changes are in a separated commit
- you don't merge immediately and wait for a comment from my part.

Regarding your changes, that's great ! :+1: 

About ``fixup ! ``, yes I know that feature, but it does'nt fit to my needs (as a reviewer). I mean : 
- for a new module PR, no problem to use fixup allways. This way, a reviewer can review all the changes, or only the last diff(s).
- but for a migration PR, as a reviewer, I like to see the all diff **of the migration** and if there are the [MIG] commit + 10 fixup commits, it's very hard to review. Migration are generally very small diff, I don't want to review all the previous unchanged code. Do you understand what I mean ?

For me, there is not perfect solution.
For the time being, AFAIK, Github does'nt provide the possiblity to see in one diff, the 10 last commits (for exemple).

regards.

**Comment by pedrobaeza** [comment]
> For the time being, AFAIK, Github does'nt provide the possiblity to see in one diff, the 10 last commits (for exemple).

Yes, it does!

Peek 12-08-2019 15-54

And GitHub also provides diff between force-pushed, so that's usually enough.

**Comment by legalsylvain** [comment]
> Yes, it does!

:+1: Gracias !

**Review by chienandalu — APPROVED** [review]
:+1: Code and functional review. Thanks!

**Comment by pedrobaeza** [comment]
/ocabot merge patch

(for bumping purchase_discount module version)

**Comment by pedrobaeza** [comment]
Still connection problems...

/ocabot merge patch

**Pull request merged**